subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c

```diff
@@ -0,0 +1,928 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include "nrf_cloud_fota.h"
+#include "nrf_cloud_mem.h"
+#include "nrf_cloud_transport.h"
+
+#include <zephyr.h>
+#include <errno.h>
+#include <stdio.h>
+#include <fcntl.h>
+#include <net/mqtt.h>
+#include <net/socket.h>
+#include <net/nrf_cloud.h>
+#include <net/fota_download.h>
+#include <net/cloud.h>
+#include <logging/log.h>
+#include <sys/util.h>
+#include <settings/settings.h>
+#include <power/reboot.h>
+
+#if defined(CONFIG_BOOTLOADER_MCUBOOT)
+#include <dfu/mcuboot.h>
+#endif
+
+#if defined(CONFIG_BSD_LIBRARY)
+#include <modem/bsdlib.h>
+#include <bsd.h>
+#endif
+
+LOG_MODULE_REGISTER(nrf_cloud_fota, CONFIG_NRF_CLOUD_LOG_LEVEL);
+
+#define TOPIC_FOTA_RCV    "/jobs/rcv"
+#define TOPIC_FOTA_REQ    "/jobs/req"
+#define TOPIC_FOTA_UPDATE "/jobs/update"
+
+#define JOB_UPDATE_MSG_TEMPLATE "[\"%s\",%d,%s]"
```

Comment:
Since we are writing our own custom thing. Why are we using JS arrays instead of CBOR?

Reply:
CBOR was not considered.  it looks like it would only save us ~10 bytes per message?  if there are compelling reasons to implement CBOR for this then it should be done.

i didn't think to use cJSON since the payloads are very simple.  i can try it out if you think it provides a benefit?

Reply:
I don't know the details of the implementation. If the cbor library is not already included in the application, I would say no since it would add to 1828 bytes, pr. update, which probably will waste the 10 bytes pr message save. An additional advantage of cbor would be parsing speed as it's lighter to parse.

If it's already included in the application, I would say go for it. Anything which can make the transport overhead lower is a great improvement.

If you can write a few generic functions with cJSON, maybe it would save a few bytes of flash.

Reply:
Remember to build with optimize for size when checking this.